util/codec/codec.go

```diff
@@ -42,9 +42,41 @@ const (
 	maxFlag          byte = 250
 )
 
+func preRealloc(b []byte, vals []types.Datum, comparable bool, hash bool) []byte {
+	var size int
+	for i := range vals {
+		switch vals[i].Kind() {
+		case types.KindInt64, types.KindUint64, types.KindMysqlEnum, types.KindMysqlSet, types.KindMysqlBit, types.KindBinaryLiteral:
+			size += sizeInt(comparable)
+		case types.KindString, types.KindBytes:
+			size += sizeBytes(vals[i].GetBytes(), comparable)
+		case types.KindMysqlTime, types.KindMysqlDuration, types.KindFloat32, types.KindFloat64:
+			size += 9
+		case types.KindNull, types.KindMinNotNull, types.KindMaxValue:
+			size += 1
+		case types.KindMysqlJSON:
+			size += 2 + len(vals[i].GetBytes())
+		case types.KindMysqlDecimal:
+			size += 1
+			if hash {
+				dec := vals[i].GetMysqlDecimal()
+				size += dec.ToHashKeySize()
+			} else {
+				size += sizeDecimal(vals[i].GetMysqlDecimal(), vals[i].Length(), vals[i].Frac())
+			}
+		default:
+			return b
+		}
+	}
+	return reallocBytes(b, size)
+}
+
 // encode will encode a datum and append it to a byte slice. If comparable is true, the encoded bytes can be sorted as it's original order.
 // If hash is true, the encoded bytes can be checked equal as it's original value.
 func encode(sc *stmtctx.StatementContext, b []byte, vals []types.Datum, comparable bool, hash bool) (_ []byte, err error) {
+	if b == nil {
```

Comment:
Does this optimization only reduce the `runtime.growslice()` overhead when the upper layer passes `nil` pointer to this function?

Reply:
previous @coocood think we should avoid calculating cost when outer give us a buffer, but now we use estimated value for decimal so we can remove this null check? @coocood